pkg/portal/middleware/aad.go

```diff
@@ -286,35 +221,63 @@ func (a *aad) callback(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	idToken, err := a.verifier.Verify(r.Context(), rawIDToken)
+	_, _, _, expiry, err := a.accessValidator.validateAccess(r.Context(), rawIDToken, a.now())
 	if err != nil {
-		a.internalServerError(w, err)
+		// we could not identify the user so we make them try to login again
+		http.Redirect(w, r, "/api/login", http.StatusTemporaryRedirect)
 		return
 	}
 
+	// only keep the cookie for the validity time of the token
+	http.SetCookie(w, &http.Cookie{
+		Name:     OIDCCookie,
+		Value:    rawIDToken,
+		Expires:  expiry,
+		SameSite: http.SameSiteStrictMode,
+		Secure:   true,
+		HttpOnly: true,
+	})
+
+	// delete the state cookie
+	http.SetCookie(w, &http.Cookie{
+		Name:   stateCookie,
+		Value:  "",
+		MaxAge: -1,
+	})
+
+	http.Redirect(w, r, "/", http.StatusTemporaryRedirect)
+}
+
+func extractInfoFromToken(token *oidc.IDToken) (preferredUsername string, groups []string, err error) {
 	var claims claims
-	err = idToken.Claims(&claims)
-	if err != nil {
-		a.internalServerError(w, err)
-		return
-	}
 
-	groupsIntersect := GroupsIntersect(a.allGroups, claims.Groups)
-	if len(groupsIntersect) == 0 {
-		http.Error(w, http.StatusText(http.StatusForbidden), http.StatusForbidden)
+	err = token.Claims(&claims)
+	if err != nil {
+		return "", nil, err
 	}
 
-	session.Values[SessionKeyUsername] = claims.PreferredUsername
-	session.Values[SessionKeyGroups] = groupsIntersect
-	session.Values[SessionKeyExpires] = a.now().Add(a.sessionTimeout).Unix()
+	return claims.PreferredUsername, claims.Groups, nil
+}
 
-	err = session.Save(r, w)
+func (d defaultAccessValidator) validateAccess(ctx context.Context, rawIDToken string, now time.Time) (forbidden bool, username string, groups []string, expiry time.Time, err error) {
+	idToken, err := d.verifier.Verify(ctx, rawIDToken)
 	if err != nil {
-		a.internalServerError(w, err)
-		return
+		return false, "", nil, time.Time{}, err
 	}
 
-	http.Redirect(w, r, "/", http.StatusTemporaryRedirect)
+	if idToken.Expiry.Before(now) {
+		// token has expired
+		return false, "", nil, idToken.Expiry, err
+	}
+	username, groups, err = extractInfoFromToken(idToken)
+	if err != nil {
+		return false, username, groups, idToken.Expiry, err
+	}
+	groupsIntersect := GroupsIntersect(d.allowedGroups, groups)
+	if len(groupsIntersect) == 0 {
+		return true, username, groups, idToken.Expiry, nil
```

Comment:
if it's not allowed, we might want to log things, so we have a trace (e.g. in case of malicious attempts of pen testing). for the other you are right, I changed it